akka-actor/src/main/scala/akka/util/MessageBuffer.scala

```diff
@@ -0,0 +1,238 @@
+
+  /**
+   * Check if the message buffer is empty.
+   *
+   * @return if the buffer is empty
+   */
+  def isEmpty: Boolean = _head eq null
+
+  /**
+   * Check if the message buffer is not empty.
+   *
+   * @return if the buffer is not empty
+   */
+  def nonEmpty: Boolean = !isEmpty
+
+  /**
+   * How many elements are in the message buffer.
+   *
+   * @return the number of elements in the message buffer
+   */
+  def size: Int = _size
+
+  /**
+   * Add one element to the end of the message buffer.
+   *
+   * @param message the message to buffer
+   * @param ref the actor to buffer
+   * @return this message buffer
+   */
+  def append(message: Any, ref: ActorRef): MessageBuffer = {
+    val node: Node = new Node(null, message, ref)
+    if (isEmpty) {
+      _head = node
+      _tail = node
+    } else {
+      _tail.next = node
+      _tail = node
+    }
+    _size += 1
+    this
+  }
+
+  /**
+   * Remove the first element of the message buffer.
+   */
+  def dropHead(): Unit = if (nonEmpty) {
+    _head = _head.next
+    _size -= 1
+    if (isEmpty)
+      _tail = null
+  }
+
+  /**
+   * Return the first element of the message buffer.
+   *
+   * @return the first element or an element containing null if the buffer is empty
+   */
+  def head(): (Any, ActorRef) =
+    if (nonEmpty) (_head.message, _head.ref)
+    else (null, null)
+
+  /**
+   * Java API
+   *
+   * Return the first element of the message buffer.
+   *
+   * @return the first element or an element containing null if the buffer is empty
+   */
+  def getHead(): akka.japi.Pair[Any, ActorRef] = {
+    import akka.japi.Pair
+    if (nonEmpty) Pair.create(_head.message, _head.ref)
+    else Pair.create(null, null)
+  }
+
+  /**
+   * Iterate over all elements of the buffer and apply a function to each element.
+   *
+   * @param f the function to apply to each element
+   */
+  def foreach(f: (Any, ActorRef) ⇒ Unit): Unit = {
+    var node = _head
+    while (node ne null) {
+      node(f)
+      node = node.next
+    }
+  }
+
+  /**
+   * Java API
+   *
+   * Iterate over all elements of the buffer and apply a function to each element.
+   *
+   * @param f the function to apply to each element
+   */
+  def forEach(f: Procedure2[Any, ActorRef]): Unit = foreach { case (message, ref) ⇒ f(message, ref) }
+}
+
+object MessageBuffer {
+  private final class Node(var next: Node, val message: Any, val ref: ActorRef) {
+    def apply(f: (Any, ActorRef) ⇒ Unit): Unit = {
+      f(message, ref)
+    }
+  }
+
+  /**
+   * Create an empty message buffer.
+   *
+   * @return an empty message buffer
+   */
+  def empty: MessageBuffer = new MessageBuffer(null, null)
+}
+
+/**
+ * A non thread safe mutable message buffer map that can be used to buffer messages inside actors.
+ *
+ * @tparam I (Id type)
+ */
+final class MessageBufferMap[I] {
+  import scala.collection.mutable
+
+  private val bufferMap = mutable.Map.empty[I, MessageBuffer]
```

Comment:
Would have expected the Java one be used I guess (a bit lighter wasn't it?)?
Not important to change right now though